### PR TITLE
Implementation of the cumulative distribution function and its inverse

### PR DESCRIPTION
**Pull request by vishwakftw**
This PR adds the `.cdf` and `.icdf` methods for distributions. @alicanb implemented these in the `Distribution` base class.

Furthermore, this PR adds these methods for `Exponential`, `Normal`, `Cauchy` and for `TransformedDistribution`. `Pareto` and `Gumbel` are converted to `TransformedDistribution` with base classes as `Exponential` and `Uniform` respectively.

Tests for checking the correctness of `.cdf` and `.icdf` methods have also been added. An autograd based check is added to check the correctness of `.cdf` and `.log_prob` as well (this doesn't depend on `TEST_NUMPY`).

Joint work with @fritzo and @alicanb . Checked and approved by @fritzo .

### CODE REVIEW

**Review by fritzo — APPROVED** [review]
LGTM Previous review: https://github.com/probtorch/pytorch/pull/122

File: test/test_distributions.py [hunk 1]
```diff
@@ -1161,6 +1161,47 @@ def test_beta_sample(self):
...
+                ]))
+
+    def test_cdf_log_prob(self):
+        # Tests if the differentiation of the CDF gives the PDF at a given value
+        for Dist, params in EXAMPLES:
+            for i, param in enumerate(params):
+                dist = Dist(**param)
+                samples = dist.sample()
+                if not samples.requires_grad:
+                    continue
```
Comment (apaszke) [inline test/test_distributions.py]: Why would an output of `sample` ever require grad?

File: test/test_distributions.py [hunk 2]
```diff
@@ -2358,17 +2413,41 @@ def setUp(self):
 
     def test_mean(self):
         for pytorch_dist, scipy_dist in self.distribution_pairs:
+            if isinstance(pytorch_dist, Cauchy):
+                continue
```
Comment (apaszke) [inline test/test_distributions.py]: A comment would be welcome (I have no idea why do we skip Cauchy)

File: test/test_distributions.py [hunk 3]
```diff
@@ -2358,17 +2413,41 @@ def setUp(self):
...
             self.assertEqual(pytorch_dist.mean, scipy_dist.mean(), allow_inf=True, message=pytorch_dist)
 
     def test_variance_stddev(self):
         for pytorch_dist, scipy_dist in self.distribution_pairs:
+            if isinstance(pytorch_dist, Cauchy):
```
Comment (apaszke) [inline test/test_distributions.py]: Same here

File: test/test_distributions.py [hunk 4]
```diff
@@ -2358,17 +2413,41 @@ def setUp(self):
...
+                continue
             if isinstance(pytorch_dist, (Multinomial, OneHotCategorical)):
                 self.assertEqual(pytorch_dist.variance, np.diag(scipy_dist.cov()), message=pytorch_dist)
                 self.assertEqual(pytorch_dist.stddev, np.diag(scipy_dist.cov()) ** 0.5, message=pytorch_dist)
             else:
                 self.assertEqual(pytorch_dist.variance, scipy_dist.var(), allow_inf=True, message=pytorch_dist)
                 self.assertEqual(pytorch_dist.stddev, scipy_dist.var() ** 0.5, message=pytorch_dist)
 
+    def test_cdf(self):
+        set_rng_seed(0)  # see Note [Randomized statistical tests]
```
Comment (apaszke) [inline test/test_distributions.py]: Is it really one of those tests? `cdf` is deterministic, so it shouldn't really matter, right? Or is it because of numerical instabilities?

File: torch/distributions/cauchy.py [hunk 5]
```diff
@@ -53,5 +53,13 @@ def log_prob(self, value):
         self._validate_log_prob_arg(value)
         return -math.log(math.pi) - self.scale.log() - (1 + ((value - self.loc) / self.scale)**2).log()
 
+    def cdf(self, value):
+        self._validate_log_prob_arg(value)
+        return torch.atan((value - self.loc) / self.scale) / math.pi + 0.5
+
+    def icdf(self, value):
+        self._validate_log_prob_arg(value)
```
Comment (apaszke) [inline torch/distributions/cauchy.py]: we might want to rename this to `_validate_sample`, since it's no longer used for args of log prob only

File: test/test_distributions.py (hunk 1 above)
Comment (fritzo) [reply test/test_distributions.py]: Good point, I think this should be
```py
samples = Variable(samples.data, requires_grad=True)
```
to make `grad()` work below.

File: test/test_distributions.py (hunk 4 above)
Comment (alicanb) [reply test/test_distributions.py]: I wanted to fix the seed just in case because of sampling, Note there is unnecessary though...

**Review by apaszke — APPROVED** [review]
